cpp/src/reductions/reduction_dispatcher.cuh

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_REDUCTION_DISPATCHER_CUH
+#define CUDF_REDUCTION_DISPATCHER_CUH
+
+#include "reduction_functions.cuh"
+
+namespace cudf {
+namespace reductions {
+
+// make an column iterator
+template<bool has_nulls, typename T_in, typename T_out, typename Op>
+auto make_iterator(const gdf_column &input, T_out identity, Op op)
+{
+    return cudf::make_iterator<has_nulls, T_in, T_out>(input, identity);
+}
+
+// make an column iterator, specialized for `cudf::reductions::ReductionSumOfSquares`
+template<bool has_nulls, typename T_in, typename T_out>
+auto make_iterator(const gdf_column &input, T_out identity,
```

Comment:
I think this design could be simplified a bit. The "Loader"s could be added back to the Reduction operators. For things like `ReductionMin`, the loader would just be the "IdentityLoader".

Then you don't need to have special overloads of `make_iterator`, you can just do:

```
    auto it_raw = cudf::make_iterator<has_nulls, T_in, T_out>(input, identity);
    return thrust::make_transform_iterator(it_raw, Op::template Loader<T_out>{});
```

Reply:
I think that  identityLoader for the default function will add extra overhead of `thrust::make_transform_iterator` since the default function just return the iterator `cudf::make_iterator` without extra `thrust::make_transform_iterator`.
If we'd like much simplified the code, I feel adding `make_iterator` into each `cudf::reduction::op::xx` like `cudf::reduction::op::mean` . The compound operators uses it.

Reply:
I'm guessing that the compiler will boil all the overhead away. But since you've got your benchmarks, you can easily test and see.

Reply:
Unfortunately I cannot easily test and see yet.
Though I've created iterator bench, I've never created reduction bench yet.
For now, I've moved `make_iterator` into `op`